src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3029,25 +3072,58 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
         public override BoundNode? VisitLocalFunctionStatement(BoundLocalFunctionStatement node)
         {
             var localFunc = node.Symbol;
+
+            // Usages state is created when we visit the function's call site or its body.
+            // In the first pass of the nullable walker, existence of usages here means that a call site has been visited.
+            // In subsequent nullable walker passes, the starting state is preserved from previous passes.
+            // In any case, existence of usages means that we have a good starting state.
+            var hasGoodStartingState = HasLocalFuncUsagesCreated(localFunc);
+
             var localFunctionState = GetOrCreateLocalFuncUsages(localFunc);
 
-            // The starting state (`state`) is the top state ("maybe null").
+            // The state for the function's body analysis starts as the top state ("maybe null").
             var state = TopState();
 
-            // Captured variables are joined with the state
-            // at all the local function use sites (`localFunctionState.StartingState`)
-            // which starts as the bottom state ("not null").
-            var startingState = localFunctionState.StartingState;
-            startingState.ForEach(
-                (slot, variables) =>
-                {
-                    var symbol = variables[variables.RootSlot(slot)].Symbol;
-                    if (Symbol.IsCaptured(symbol, localFunc))
+            if (!hasGoodStartingState)
+            {
+                // For unreachable local functions, top-level captured variables are set to "not null",
+                // ignoring nested slots to avoid depending on slot allocation order
+                // (e.g., whether we have seen a class field or not already).
+                state.Normalize(this, _variables);
+                state.ForEach(
+                    (slot, variables) =>
                     {
-                        SetState(ref state, slot, GetState(ref startingState, slot));
-                    }
-                },
-                _variables);
+                        if (variables[slot].ContainingSlot == 0) // top-level variables only
```

Comment:
>if (variables[slot].ContainingSlot == 0) // top-level variables only [](http://example.com/codeflow?start=24&length=68)

Why is this check still necessary? It doesn't look like it is necessary in the `else` part of the enclosing `if`. Is anything going to change if we remove it. The way it looks to me, both branches should be setting the state for the same slots, just use different states.

Reply:
Are you saying that ```IsCaptured``` can return true for a nested slot?

Reply:
Yes, if `slot` is nested, `variables[variables.RootSlot(slot)].Symbol` returns the root (not nested) symbol which is what's then checked via `IsCaptured`.

Reply:
I guess we could check `IsCaptured` of the nested symbol instead, but the `variables[slot].ContainingSlot == 0` check looks much faster.

Reply:
Sorry, I am confused. It is yes or no question. Can ```IsCaptured``` return true for a nested slot?

Reply:
> Can `IsCaptured` return true for a nested slot?

No.

Reply:
> No

Then I think this if condition is unnecessary, and we  should simply rely on ```IsCaptured``` doing its job. It looks to me that it filters out everything that is not a local or parameter pretty early.